### ISSUE TITLE
Zed update fails to restart and spawn endless lsof process instead

### ISSUE THREAD

**Issue by afossa** [zed-industries/zed#15689]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Open the Zed menu, choose `Check for updates` and follow the instructions. After the update is completed, click on `Restart Zed` to load the updated version.
At this point, the Zed window closes correctly, but instead of reopening on the new version, an `lsof` process that uses 100% of a CPU core is spawn. This process ends only after a logout and re-login.
After re-login, the new version works as expected.
Zed previews installed both with the recommended bash script and by manually downloading/unpacking the tarball are affected.

### Environment

Zed: v0.147.2 (Zed Preview)
OS: Linux X11 debian trixie/sid
Kernel: 6.9.12-amd64
Memory: 11.6 GiB
Architecture: x86_64
GPU: NVIDIA GeForce GTX 950M || NVIDIA || 535.183.01

### If applicable, add mockups / screenshots to help explain present your vision of the feature

![Screenshot_20240802_172658](https://github.com/user-attachments/assets/a67f46bf-1087-42cc-8c0a-2ebf0ec5b52e)

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

**Comment by notpeter** [comment]
I think this is likely a duplicate of:
- https://github.com/zed-industries/zed/issues/14291

And perhaps fixed by:
- https://github.com/zed-industries/zed/pull/16797

Can you try manually updating (re-run the installer) to the newest Zed Preview.
As Zed Preview v0.151.0-pre included the fix from #16797

If it happens again after that, please comment and I'm happy to reopen.

**Comment by afossa** [comment]
I tried the release version after this, and the issue is still present in [v0.151.1](https://github.com/zed-industries/zed/tree/v0.151.1). This version should include the fix you mentioned [as you can see](https://github.com/zed-industries/zed/blob/v0.151.1/crates/gpui/src/platform/linux/platform.rs).

This issue does not appear when installing Zed from scratch (either release or preview). To reproduce it, you need Zed already installed, and wait until an update is available. Then open Zed and wait until the update is downloaded and installed. At this point, follow the prompt in the bottom left of the window to relaunch Zed on the new version. Here is when Zed never opens again and the process runs indefinitely.

**Comment by JosephTLyons** [comment]
This issue is missing core labels. Adding new issue labels for triage.

**Comment by notpeter** [comment]
@afossa Are you still able to reproduce this issue?

This is the location in the code where `lsof` is spawned.   
https://github.com/zed-industries/zed/blob/06533d5a686b86388ba49feb26c3f1930fdeffa8/crates/gpui/src/platform/linux/platform.rs#L180-L198

```
lsof -nP -iTCP -a -p {pid} 
```

Are there anything distinct about your system which you could imagine causing lsof like this to hang? (e.g. running under SELinux or with restricted capabilities, extremely large numbers of open ports/files, do you have `lsof` aliased or as a non standard binary).  If you run this command in an another terminal while Zed is running does it hang?

```
lsof -nP -iTCP -a -p $(pidof zed)
```

**Comment by majutsushi** [comment]
I have the same issue, and there's not really anything special about my system that I can think of. It's a standard Debian Sid installation.

I noticed that when this situation occurs, there are actually two `lsof`s running:
```
$ psgrep lsof
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
jan       879571  0.0  0.0   8640  3404 ?        S    15:44   0:00 /bin/bash -c              while kill -0 867426 2>/dev/null; do                 sleep 0.1             done              while lsof -nP -iTCP -a -p 867426 2>/dev/null; do                 sleep 0.1             done              /home/jan/.local/zed.app/libexec/zed-editor
jan       879575  0.0  0.0   9696  3208 ?        S    15:44   0:00 lsof -nP -iTCP -a -p 867426
jan       879576 99.9  0.0   9696   584 ?        R    15:44   0:33 lsof -nP -iTCP -a -p 867426
```

Interestingly the second one, which is the one running at 100% CPU, is getting replaced with another lsof process after a few minutes which also runs at 100% CPU. After a few repeats of this both processes die and Zed starts up normally.

When I attach strace to the second lsof process I see a constant stream of this:
```
15:49:16.640700 close(848299962)        = -1 EBADF (Bad file descriptor)
15:49:16.640738 close(848299963)        = -1 EBADF (Bad file descriptor)
15:49:16.640776 close(848299964)        = -1 EBADF (Bad file descriptor)
15:49:16.640814 close(848299965)        = -1 EBADF (Bad file descriptor)
15:49:16.640851 close(848299966)        = -1 EBADF (Bad file descriptor)
15:49:16.640889 close(848299967)        = -1 EBADF (Bad file descriptor)
15:49:16.640926 close(848299968)        = -1 EBADF (Bad file descriptor)
15:49:16.640964 close(848299969)        = -1 EBADF (Bad file descriptor)
```

It seems to be iterating over all possible file descriptors for some reason, and then quits when it runs out of them. No idea why the parent lsof would restart it though.

Running the lsof command manually immediately exits as expected.

I suspect that this may be some sort of race condition, where running lsof immediately after the `kill -0` fails runs into some not-fully-cleaned-up state that then confuses lsof.

**Comment by afossa** [comment]
It seems fixed now. I am on Ubuntu 22.04 and I do not have any special configuration.

**Comment by notpeter** [comment]
Glad to hear. If you run into again and I'm happy to reopen.

I wonder if it's a race condition of some sort where lsof gets stuck trying to read something from /proc that disappears. Maybe a "sleep 0.1" or some such.
